stl/inc/xprint.h

```diff
@@ -0,0 +1,77 @@
+// xprint.h internal header (core)
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _XPRINT_H
+#define _XPRINT_H
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#include <xfilesystem_abi.h>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_EXTERN_C
+
+enum class __std_file_stream_pointer : uintptr_t { _Invalid = 0 };
+enum class __std_unicode_console_handle : intptr_t { _Invalid = -1 };
+
+struct __std_unicode_console_retrieval_result {
+    __std_unicode_console_handle _Console_handle;
+
+    // For this, we have a few potential return values:
+    //
+    //   - __std_win_error::_Success: The operation completed successfully. This is the only value for which the
+    //     _Console_handle field has a well-defined value.
+    //
+    //   - __std_win_error::_File_not_found: The FILE* provided is valid, but it is determined to not be associated
+    //     with a unicode console. In this case, printing should fall back to vprint_nonunicode().
+    //
+    //   - __std_win_error::_Not_supported: The FILE* provided does not actually have an associated output stream. In
+    //     this case, the entire print can safely be elided, thanks to the "as-if" rule.
+    //
+    //   - __std_win_error::_Invalid_parameter: The FILE* provided is invalid. A std::system_error exception should be
+    //     thrown if this value is returned within the FILE* overload of vprint_unicode().
+    __std_win_error _Error;
+};
+
+_NODISCARD _Success_(return._Error == __std_win_error::_Success) __std_unicode_console_retrieval_result
+    __stdcall __std_get_unicode_console_handle_from_file_stream(_In_ const __std_file_stream_pointer _Stream) noexcept;
+
+_NODISCARD _Success_(return == __std_win_error::_Success) __std_win_error
+    __stdcall __std_print_to_unicode_console(_In_ const __std_unicode_console_handle _Console_handle,
+        _In_ const char* const _Str, _In_ const unsigned long long _Str_size) noexcept;
+
+_END_EXTERN_C
+
+_STD_BEGIN
+
+inline constexpr bool _Is_ordinary_literal_encoding_utf8 = []() {
+// For the MSVC, we use the _MSVC_EXECUTION_CHARACTER_SET to get the ordinary literal encoding
+// exactly.
+//
+// For Clang, we use the hack suggested in P2093R14. Ideally, we would use a better solution.

```

Comment:
Clang 15 provides `_MSVC_EXECUTION_CHARACTER_SET`, so it doesn't need this hack. See also [`_Is_execution_charset_self_synchronizing` in `<format>`](https://github.com/microsoft/STL/blob/8ddf4da/stl/inc/format#L828) for prior art.

Note that EDG (which powers IntelliSense) doesn't support `_MSVC_EXECUTION_CHARACTER_SET`, so you need to make sure that the code is valid when the macro is not defined, but it doesn't matter which value is used in this case.

Reply:
Ah, so I see. I suppose I incorrectly assumed that the macro was an MSVC-specific thing due to its name and the relative lack of documentation for it.